fl4health/losses/mkmmd_loss.py

```diff
@@ -0,0 +1,288 @@
+# multi-kernel maximum mean discrepancy
+# cao bin, HKUST, China, binjacobcao@gmail.com
+# free to charge for academic communication
+
+from logging import INFO
+from typing import Optional
+
+import torch
+from flwr.common.logger import log
+from qpth.qp import QPFunction, QPSolvers
+
+
+class MkMmdLoss(torch.nn.Module):
+    def __init__(
+        self,
+        device: torch.device,
+        gammas: Optional[torch.Tensor] = None,
+        betas: Optional[torch.Tensor] = None,
+        minimize_type_two_error: bool = True,
+    ) -> None:
+        """
+        Compute the multi-kernel maximum mean discrepancy (MK-MMD) between the source and target domains. Also allows
+        for optimization of the coefficients, beta
+
+        Args:
+            device (torch.device): Device onto which tensors should be moved
+            gammas (Optional[torch.Tensor], optional): These are known as the length-scales of the RBF functions used
+                to compute the Mk-MMD distances. The length of this list defines the number of kernels used in the
+                norm measurement. If none, a default of 29 kernels is used. Defaults to None.
+            betas (Optional[torch.Tensor], optional): These are the linear coefficients used on the basis of kernels
+                to compute the Mk-MMD measure. If not provided, a unit-length, random default is constructed. These
+                can be optimized using the functions of this class. Defaults to None.
+            minimize_type_two_error (Optional[bool], optional): Whether we're aiming to minimize the type II error in
+                optimizing the betas or maximize it. The first coincides with trying to minimize feature distance. The
+                second coincides with trying to maximize their feature distance. Defaults to True.
+        """
+        super().__init__()
+        self.device = device
+        if gammas is None:
+            # Note arange is not inclusive, so this ends up being [-3.5, 1] in steps of 0.5
+            default_gamma_powers = torch.arange(-3.5, 1.25, 0.25, device=device)
+            self.gammas = torch.pow(2.0, default_gamma_powers)
+        else:
+            self.gammas = gammas
+        self.kernel_num = len(self.gammas)
+
+        if betas is None:
+            rand_coefficients = torch.rand((self.kernel_num, 1)).to(self.device)
+            # normalize the coefficients to sum to 1
+            self.betas = (1 / torch.sum(rand_coefficients)) * rand_coefficients
+        else:
+            assert betas.shape == (self.kernel_num, 1)
```

Comment:
Perhaps we can also sanity check that the entered betas sum to 1?

Reply:
I figured out in one of tests fails in that assertion error, should I change tests? Also is it better to raise error when preset betas are not sum to 1 or just normalize before setting betas?

Reply:
It also fails when we set random betas as they don't sum up exactly to 1 due to computational approximation.

Reply:
For the test, let's just set the beta values explicitly after the object is created to get around the assertion without having to re-compute the test? A bit of cheating, but I think that's fine.

For the numerical fluctuations, we can just do an approximate assertion, i.e. $\vert \sum betas - 1 \vert < 0.0001$ or something.